pyxem/generators/orientation_refinement_generator.py

```diff
@@ -0,0 +1,205 @@
+        method: str, optional
+            Minimization method to use, should be one of 'nelder', 'powell', 'cobyla', 'least-squares'
+        fit_tol: float
+            Tolerance for termination. For detailed control, use solver-specific options.
+        """
+        if not projector:
+            projector = self.projector
+
+        f_kws = kwargs.get("kws", None)
+        
+        def objfunc(params, img):
+            cx = params["center_x"].value
+            cy = params["center_y"].value
+            al = params["alpha"].value
+            be = params["beta"].value
+            ga = params["gamma"].value
+            sc = params["scale"].value
+            
+            proj = projector.get_projection(al, be, ga)
+            pks = proj[:,3:6]
+            score = get_score_shape(img, pks, sc, cx, cy)
+
+            return 1e3/(1+score)
+
+        params = lmfit.Parameters()
+        params.add("center_x", value=result.center_x, vary=vary_center, min=result.center_x - 2.0, max=result.center_x + 2.0)
+        params.add("center_y", value=result.center_y, vary=vary_center, min=result.center_y - 2.0, max=result.center_y + 2.0)
+        params.add("alpha", value=result.alpha, vary=vary_alphabeta)
+        params.add("beta",  value=result.beta,  vary=vary_alphabeta)
+        params.add("gamma", value=result.gamma, vary=vary_gamma)
+        params.add("scale", value=result.scale, vary=vary_scale, min=result.scale*0.8, max=result.scale*1.2)
+        
+        args = img,
+
+        res = lmfit.minimize(objfunc, params, args=args, method=method, tol=fit_tol, kws=f_kws)
+
+        if verbose:
+            lmfit.report_fit(res)
+                
+        p = res.params
+        
+        alpha, beta, gamma = [round(p[key].value, 4) for key in ("alpha", "beta", "gamma")]
+        scale, center_x, center_y = [round(p[key].value, 2) for key in ("scale", "center_x", "center_y")]
+        
+        proj = projector.get_projection(alpha, beta, gamma)
+        pks = proj[:,3:6]
+        
+        score = round(get_score_shape(img, pks, scale, center_x, center_y), 2)
+        
+        # print "Score: {} -> {}".format(int(score), int(score))
+        
+        refined = IndexingResult(score=score,
+                                 number=result.number,
+                                 alpha=alpha,
+                                 beta=beta,
+                                 gamma=gamma,
+                                 center_x=center_x,
+                                 center_y=center_y,
+                                 scale=scale,
+                                 phase=result.phase)
+        
+        return refined
+
+    def probability_distribution(self, img, result, projector=None, verbose=True, vary_center=False, vary_scale=True):
+        """https://lmfit.github.io/lmfit-py/fitting.html#lmfit.minimizer.Minimizer.emcee
+
+        Calculate posterior probability distribution of parameters"""
+        import corner
+        import emcee
+
```

Comment:
What are these packages? If they're stable and well supported we can depend on them if needs be, but will need to be added to requirements in setup.py, appveyor.yml and .travis.yml